igvc_platform/src/motor_controller/main.cpp

```diff
@@ -1,247 +1,286 @@
-          ROS_INFO("Successfully set I values");
-        }
-        else if (ret.at(1) != 'V')
-        {
-          ROS_ERROR_STREAM("Recieved unknown string while setting PID values " << ret);
-        }
-      }
-      else if (ret.empty())
-      {
-        ROS_ERROR_STREAM("Empty return from motor board while sending PID values.\t" << ret);
-      }
-    }
-    rate.sleep();
-  }
-  ROS_INFO_STREAM("Sucessfully set all PID values");
+  setPID(sock, rate); // Set PID Values on mbed
 
-  // sends down motor commands and recieves multiple responses back
-  while (ros::ok() && port.isOpen())
+  // send motor commands
+  while (ros::ok())
   {
-    std::string msg = "$" + (enabled ? toBoundedString(current_motor_command.left_velocity) : toBoundedString(0.0)) +
-                      "," + (enabled ? toBoundedString(current_motor_command.right_velocity) : toBoundedString(0.0)) +
-                      "\n";
+      /* This is the buffer where we will store the request message. */
+      uint8_t requestbuffer[256];
+      size_t message_length;
+      bool status;
 
-    port.write(msg);
+      /* allocate space for the request message to the server */
+      RequestMessage request = RequestMessage_init_zero;
 
-    std::string ret = port.readln();
-    size_t dollar = ret.find('$');
-    size_t pound = ret.find('#');
-    int count = 0;
-    while ((dollar != std::string::npos || pound != std::string::npos) && count <= messages_to_read)
-    {
-      count++;
-      // size_t end = ret.find('\n');
-      size_t end = ret.size();
-      std::vector<std::string> tokens;
+      /* Create a stream that will write to our buffer. */
+      pb_ostream_t ostream = pb_ostream_from_buffer(requestbuffer, sizeof(requestbuffer));
+
+      /* indicate that speed fields will contain values */
+      request.has_speed_l = true;
+      request.has_speed_r = true;
+
+      /* fill in the message fields */
+      request.speed_l = static_cast<float>(current_motor_command.left_velocity);
+      request.speed_r = static_cast<float>(current_motor_command.right_velocity);
 
-      if (ret.size() <= 0)
+      /* encode the protobuffer */
+      status = pb_encode(&ostream, RequestMessage_fields, &request);
+      message_length = ostream.bytes_written;
+
+      /* check for any errors.. */
+      if (!status)
       {
-        ROS_INFO_STREAM("Invalid number of tokens from motor board, ret: " << ret);
+          ROS_ERROR_STREAM("Encoding failed: " << PB_GET_ERROR(&ostream));
+          ros::shutdown();
       }
-      else if (pound != std::string::npos)
+
+      /* Send the message strapped to a pigeon's leg! */
+      sock.sendMessage(reinterpret_cast<char*>(requestbuffer), message_length);
+
+      /* Read response from the server */
+      size_t n; // n is the response from socket: 0 means connection closed, otherwise n = num bytes read
+      uint8_t buffer[256]; // buffer to read response into
+
+      memset(buffer, 0, sizeof(buffer));
```

Comment:
I think `uint8_t buffer[256] = {0};` is cleaner than an explicit memset.

Reply:
I think i'll leave it as `memset` as I find it easier to understand in this case